### ISSUE TITLE
Correct use of custom `filter_spec`

### ISSUE THREAD

**Issue by tomsturges** [patrick-kidger/equinox#1176]
Please may I ask a question to try and understand the correct way to use equinox to build my own trainable models.
I am using a package called `Chromatix` for simulating wave optics.
In the [examples](https://chromatix.readthedocs.io/en/latest/training/#3-using-opticalsystem-and-equinox-partitioning-to-choose-what-parameters-are-optimizable) they create some trainable `model`, then create a custom `filter_spec` using `eqx.tree_at` to specify what they want to be trainable. Then they write:

```
@jax.jit
def update(model, opt_state, target):
    parameters, state = eqx.partition(model, filter_spec)
    ....
```

This confuses me, because it seems to go against the [pattern](https://docs.kidger.site/equinox/all-of-equinox/#2-filtering) on the Filtering section of the All of Equinox docs, where either you:

1. specify your `filter_spec`, split the `model`, use a function that takes these split parts separately, and specify that the `static` part is static using `static_argnums=1`, as below

```
filter_spec = eqx.is_array # or something more complicated
params, static = eqx.partition(model, filter_spec)
@ft.partial(jax.jit, static_argnums=1) 
@jax.grad
    def loss(params, static, x, y):
    ...
```

or,
2. if you don't want to use a custom `filter_spec` and are happy to train all arrays then you can use the helpers `eqx.filter_jit` and
`eqx.filter_grad`

If I could help understand why (or why not) what Chromatic does is correct, then I hope that should help me understand how to structure my code. Thanks for any help.

**Comment by lockwo** [comment]
Looking through their example, my understanding is that they are using filtering in a reasonable, albeit different, manner. The filtering docs are mostly about filtering out non-JAX types (i.e. so you don't get an error when using jax.jit). In their example, they have a pytree of arrays, but only want to differentiate with respect to some of them. So they filter out the ones they don't want, then pass that pytree to a jax.grad function (then combine). That way jax.grad sees Nones and doesn't differentiate.

**Comment by tomsturges** [comment]
Ah ok now I think I get it. `eqx.filter_jit` filters out all non-arrays so that `jax.jit` works, sort of like a beefed up version of `static_argnums=...` in `jax.jit`. However, Chromatix doesn't need this because that update function is already `jax.jit` friendly. 

In a similar way, `eqx.filter_grad` also filters out non-arrays, but the goal here is to specify what to take gradients of, sort of like a beefed up version of `argnums=...` from `jax.grad`. But if you want even more customisation of what to take gradients of then you can use `eqx.partition`.

I had the wrong mental model that _either_ you use the `eqx.filter_{jit, grad, ...}` helpers _OR_ you use `eqx.partition` together with `jax.{jit, grad, ...}`, but now I see how you can use a combination. 

Ok thanks!
